## EXPLORING EBAY CAR SALES DATA

- The aim of this project is to clean the data and analyze the included used car listing

In [1]:
import numpy as np
import pandas as pd
autos = pd.read_csv('autos.csv', encoding="Windows-1252")
print(autos.head(0))

Empty DataFrame
Columns: [dateCrawled, name, seller, offerType, price, abtest, vehicleType, yearOfRegistration, gearbox, powerPS, model, odometer, monthOfRegistration, fuelType, brand, notRepairedDamage, dateCreated, nrOfPictures, postalCode, lastSeen]
Index: []


In [2]:
print(autos.head())
print('\n')
print(autos.info())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

**- dateCrawled** - When this ad was first crawled. All field-values are taken from this date.

**- name** - Name of the car.

**- seller** - Whether the seller is private or a dealer.

**- offerType** - The type of listing

**- price** - The price on the ad to sell the car.

**- abtest** - Whether the listing is included in an A/B test.

**- vehicleType** - The vehicle Type.

**- yearOfRegistration** - The year in which which year the car was first registered.

**- gearbox** - The transmission type.

**- powerPS** - The power of the car in PS.

**- model** - The car model name.

**- kilometer** - How many kilometers the car has driven.

**- monthOfRegistration** - The month in which which year the car was first registered.

**- fuelType** - What type of fuel the car uses.

**- brand** - The brand of the car.

**- notRepairedDamage** - If the car has a damage which is not yet repaired.

**- dateCreated** - The date on which the eBay listing was created.

**- nrOfPictures** - The number of pictures in the ad.

**- postalCode** - The postal code for the location of the vehicle.

**- lastSeenOnline** - When the crawler saw this ad last online.

- The aim of this project is to clean the data and analyze the included used car listings.

- autos is a dataset with incorrect and inconsistent column labels i.e mix of small and capital letters.
- The columns labels has to be converted to all lower case and snake case for easy analysis
- The dataset has 50000 rows and 20 columns, some of the columns has null values

In [3]:
autos.columns = ["date_crawled","name","seller","offer_type","price","abtest","vehicle_type",
"registration_year",
"gearbox",
"power_ps",
"model",
"odometer",
"registration_month",
"fuel_type",
"brand",
"unrepaired_damage",
"ad_created",
"no_of_pictures","postal_code",
 "last_seen"]
print(autos.columns)



Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [4]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

- The columns names was changed from a camelcase to a snake case for easy exploration

In [5]:
autos.describe(include="all")
print(autos["offer_type"].value_counts())

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64


- The seller(49999 private and 1 public) and offer_type(49999 angebot and 1 Gesuch) columns have mostly one value and are therefore not significant in the analysis.
- The columns vehicle_type,gearbox,model,fueltype and unrepaired damage needs more investigation as they contain null values and some unrealistic values
- The price and odometer columns are numeric data that are stored as text and need to be cleaned


In [6]:
autos['price'] = autos['price'].str.replace("$","").str.replace(",","").astype(int)
unique_price = autos['price'].head()
print(unique_price)

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32


In [7]:
autos.rename({"price":"price_dollars"},axis=1,inplace=True)

- The price column has been converted to numeric data i.e the dollar and comma has been removed and coverted to integer.
- The column name has been changed to price_dollars to indicate that the price is in (dollars).

In [8]:
autos['odometer'] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)
unique_odometer = autos["odometer"].head()
print(unique_odometer)

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int32


In [9]:
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

In [10]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_dollars',
       'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps',
       'model', 'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


- The odometer column has been converted to numeric data(all non-numeric characters has been removed and the column name has been changed to odometer_km.

In [11]:
print(autos["price_dollars"].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64


In [12]:
print(autos["odometer_km"].describe())

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [13]:
print(autos["price_dollars"].unique().shape)

(2357,)


In [14]:
print(autos["odometer_km"].unique().shape)

(13,)


In [15]:
print(autos["price_dollars"].value_counts().head().sort_index(ascending=False).head())

2500     643
1500     734
1000     639
500      781
0       1421
Name: price_dollars, dtype: int64


In [16]:
print(autos["price_dollars"].max())
print(autos["price_dollars"].min())

99999999
0


In [17]:
print(autos["odometer_km"].max())
print(autos["odometer_km"].min())

150000
5000


In [18]:
print(autos["price_dollars"].value_counts().head().sort_values(ascending=False).head())

0       1421
500      781
1500     734
2500     643
1000     639
Name: price_dollars, dtype: int64


- The odometer of the cars in the dataset ranges from 5000 to 150000 which is quite reasonable for auctions cars.
- The price of the cars ranges from 0 to 99,999,999. the price column need to be cleaned further as there cant be a car with the price 0 dollars and an action car can't be as high as 99,999,999.
- The outliers are to be removed.

In [19]:
autos = autos[autos["price_dollars"].between(1,700000)]
print(autos["price_dollars"].describe())

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_dollars, dtype: float64


- The outliers on the price column has been removed. The range between the 1 dollars 700000 dollars was used because the most expensive car on eBay in 2016 is 649000 dollars and 1 dollars is sometimes used to attract bidders.

In [20]:
print(autos[['date_crawled','ad_created','last_seen']][:5])
print()
print(autos['date_crawled'].isnull().sum())

          date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

0


In [21]:
crawled_date_only = (autos["date_crawled"].str[:10])
ad_date_only = (autos["ad_created"].str[:10])
last_seen_date_only = (autos["last_seen"].str[:10])
print(crawled_date_only.head())
print(ad_date_only.head())
print(last_seen_date_only.head())

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object
0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: ad_created, dtype: object
0    2016-04-06
1    2016-04-06
2    2016-04-06
3    2016-03-15
4    2016-04-01
Name: last_seen, dtype: object


In [22]:
print(crawled_date_only.value_counts(normalize=True,dropna=False).sort_index(ascending=True))
print(ad_date_only.value_counts(normalize=True,dropna=False).sort_index(ascending=True))
print(last_seen_date_only.value_counts(normalize=True,dropna=False).sort_index(ascending=True))

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858

######  CRAWLED DATE, ADVERT CREATION DATE AND LAST SEEN DATE
- The first car was crawled on 5th of march 2016 and the last one on the 7th of april 2016.The crawled date column 
- The first advert on the dataset was created on 11th of june 2015 and the last advert is on the 7th of april 2016


In [23]:
print(crawled_date_only.notnull().sum())

48565


In [24]:
print(ad_date_only.isnull().sum())
print(last_seen_date_only.isnull().sum())

0
0


In [25]:
print(autos['registration_year'].describe())

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


- The regisration year contains 9999 as the maximum and 1000 as the minimum. This column is about the registration year of the advertised vehicle.There is no vehicle registered in 9999 as this is just 2020 and a car registered in 1000 when car is yet to be invented.

A car can't be registered after the advertisement.The latest advertisement year is 2016 and therefore any registration date above 2016 is inaccurate.
The earliest registration date can be taken to be 1900. therefore we will use the rows the the registration year range between 1900 and 2016.

In [26]:
autos = (autos[autos["registration_year"].between(1900,2016)])
print(autos["registration_year"].value_counts(normalize=True).sort_index(ascending=False).head())

2016    0.026135
2015    0.008397
2014    0.014203
2013    0.017202
2012    0.028063
Name: registration_year, dtype: float64


In [27]:
print(autos['registration_year'].shape)

(46681,)


##### EXPLORING PRICE BY BRANDS

In [28]:
print(autos['brand'].value_counts(normalize=True))

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

In [29]:
print(autos["brand"].value_counts(normalize=True).head())

volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
Name: brand, dtype: float64


German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [38]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brand = brand_counts[brand_counts > .05]
common_brands = common_brand.index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [31]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price_dollars"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices


{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

Of the top 5 brands, there is a distinct price gap:

Audi, BMW and Mercedes Benz are more expensive

Ford and Opel are less expensive

Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

In [32]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [33]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [34]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [35]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.